e2e-tests/src/driver.rs

```diff
@@ -97,131 +97,113 @@ impl Driver {
-
-        while let Some(entry) = entries.next_entry().await? {
-            let path = entry.path();
-            if path.is_dir() {
-                let test_file_path = path.join("test.json");
+        let protocols_dir = self.environment.input_dir.join("protocols");
+
+        for protocol_name in &self.environment.protocols {
+            let protocol_name = protocol_name.to_string().to_lowercase();
+
+            if initialized_protocols.contains_key(&protocol_name) {
+                continue;
+            }
+
+            for sandbox_cfg in &self.config.protocol_sandboxes {
+                let config_protocol_name = match sandbox_cfg {
+                    ProtocolSandboxConfig::Stellar(_) => "stellar",
+                    ProtocolSandboxConfig::Near(_) => "near",
+                    ProtocolSandboxConfig::Icp(_) => "icp",
+                    ProtocolSandboxConfig::Ethereum(_) => "ethereum",
+                };
+
+                if config_protocol_name == protocol_name {
+                    let sandbox_env = match sandbox_cfg {
+                        ProtocolSandboxConfig::Stellar(config) => {
+                            ProtocolSandboxEnvironment::Stellar(StellarSandboxEnvironment::init(
+                                config.clone(),
+                            )?)
+                        }
+                        ProtocolSandboxConfig::Near(config) => ProtocolSandboxEnvironment::Near(
+                            NearSandboxEnvironment::init(config.clone()).await?,
+                        ),
+                        ProtocolSandboxConfig::Icp(config) => ProtocolSandboxEnvironment::Icp(
+                            IcpSandboxEnvironment::init(config.clone())?,
+                        ),
+                        ProtocolSandboxConfig::Ethereum(config) => {
+                            ProtocolSandboxEnvironment::Ethereum(EthereumSandboxEnvironment::init(
+                                config.clone(),
+                            )?)
+                        }
+                    };
+
+                    initialized_protocols.insert(protocol_name.clone(), sandbox_env);
+                    break;
+                }
+            }
+        }
+
+        for protocol_name in &self.environment.protocols {
+            let protocol_name = protocol_name.to_string().to_lowercase();
+            let protocol_path = protocols_dir.join(&protocol_name);
+
+            if !protocol_path.exists() || !protocol_path.is_dir() {
+                self.environment
+                    .output_writer
+                    .write_str(&format!("No directory for protocol: {}", protocol_name));
+                continue;
+            }
+
+            let Some(sandbox) = initialized_protocols.get(&protocol_name) else {
+                bail!("Sandbox not initialized for protocol: {}", protocol_name);
+            };
+
+            let mero = self.setup_mero(&vec![sandbox.clone()]).await?;
```

Comment:
1. you can change `setup_mero` to accept a `&[_]` instead of a `&Vec<_>`

    ```suggestion
                let mero = self.setup_mero(&[sandbox.clone()]).await?;
    ```

    and finally you won't even need to clone if you took in a `&[impl AsRef<ProtocolSandboxEnvironment>]`

2. won't this mean we'll only ever have nodes with 1 protocol configuration? even though we can specify `--protocols near,ethereum`, it'll only setup `node1` with `near`, and `node2` with `ethereum`, forcing `node1` to use the relayer when it needs to do any ethereum operations and vice-versa, correct me if I'm wrong

cc: @fbozic

Reply:
this is unresolved